### ISSUE TITLE
API issue

### ISSUE THREAD

**Issue by w1tw0lf** [swartjean/ha-eskom-loadshedding#10]
Getting the follow error this morning

Logger: custom_components.eskom_loadshedding
Source: helpers/update_coordinator.py:212
Integration: Eskom Loadshedding Interface (documentation, issues)
First occurred: 7:58:21 AM (10 occurrences)
Last logged: 8:12:22 AM

Error fetching eskom_loadshedding data: Error, invalid loadshedding stage received from API after 50 attempts

Integration: v1.0.3

**Comment by swartjean** [comment]
Hi @w1tw0lf,

The Eskom loadshedding API is not a particularly well-polished machine. It looks like they may be having some issues this morning.

The `GetStatus` endpoint of their API sometimes returns a stage of `-1`, which has no meaning as far as I can tell. In the event that I receive this response, I simply retry until I get something valid. If 50 attempts elapse with no valid response, this exception gets thrown.

I seem to be getting exclusively `-1` responses if I manually query their API this morning. Maybe you could confirm:

https://loadshedding.eskom.co.za/LoadShedding/GetStatus

It also looks like their main loadshedding website is having some issues:

https://loadshedding.eskom.co.za/loadshedding/Index

In short, looks like an Eskom problem. Let's keep an eye on it over the rest of the day and then I will close this issue. Unfortunately there is not much I can do on the integration side to resolve it.

**Comment by w1tw0lf** [comment]
Yup, seems like the usual eskom problems... 😁 

Just thought i might be with the new update of home assistant that was released this am.

**Comment by Surgical71** [comment]
I also thought it might be related to my HA core that I updated. But everything was working fine last night after the update so this is just the usual Eskom issue

**Comment by swartjean** [comment]
Thanks for confirming guys 👍 

I am going to close this issue because there is nothing that we can do to address it on the integration side. I'll keep an eye on it but hopefully Eskom sorts themselves out soon.

**Comment by 0xa51f** [comment]
Hi @swartjean  - with loadshedding now resumed, it appears that ESKOM have fixed their API?

i get the correct result at: https://loadshedding.eskom.co.za/LoadShedding/GetStatus

However, I am having huge delays with startup since v1.0.4 (was quicker on v1.0.3 i think)

I am running core-2021.5.4

see logs and note timestamps, during which, my HA is not available whislt this integration tries to complete setup:

2021-05-17 10:54:12 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: eskom_loadshedding
2021-05-17 10:55:12 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: eskom_loadshedding
2021-05-17 10:56:12 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: eskom_loadshedding
2021-05-17 10:57:12 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: eskom_loadshedding
2021-05-17 10:58:12 WARNING (MainThread) [homeassistant.bootstrap] Setup timed out for stage 2 - moving forward

**Comment by swartjean** [comment]
Thanks @asifkassam, you're right that the Eskom API does seem to be happier again now. 👍 

I noticed the startup issue on my own server - I'm pretty sure it is just the integration running through all of the attempts to the API. I'm going to check it out this evening and I will probably drop the number of attempts significantly and release a new build.

I'm going to reopen this issue in order to track the slow startup problem even though it is slightly different from its original purpose.

**Comment by 0xa51f** [comment]
thanks @swartjean - i have deleted the integration, restarted, then re-installed it and the problem appears to have resolved itself.

I will let you know if it starts to happen again

**Comment by swartjean** [comment]
I suspect the only reason that the integration loaded quickly now is because the Eskom API is working properly. If the API starts reporting `-1` again then the startup issue will return.

But thank you for confirming regardless 👍

**Comment by swartjean** [comment]
I have significantly reduced the number of API attempts that need to fail before returning a stage of `0`. This should improve the startup delay when the API is returning a value of `-1` again.

I will close this as soon as I can test after loadshedding ends on Tuesday.

**Comment by 0xa51f** [comment]
> I have significantly reduced the number of API attempts that need to fail before returning a stage of `0`. This should improve the startup delay when the API is returning a value of `-1` again.
> 
> I will close this as soon as I can test after loadshedding ends on Tuesday.

Still taking a significant time to start, even though https://loadshedding.eskom.co.za/LoadShedding/GetStatus returns a valid response (1 - no loadshedding?) quickly

![Screenshot_20210519-095012_Home Assistant](https://user-images.githubusercontent.com/67918736/118775993-c8916680-b887-11eb-94f2-a64a5b6fc9c5.png)

**Comment by fyford** [comment]
Is the API down again?

**Comment by swartjean** [comment]
Closing this, the integration no longer uses the Eskom API.
